File: CSMobileBase/Classes/ApexSyncDownTarget.m
//
//  ApexSyncDownTarget.m
//  CSMobileBase
//
//  Created by Mayank Bhayana on 10/01/19.
//

#import "ApexSyncDownTarget.h"

@implementation ApexSyncDownTarget

- (instancetype)initWithDict:(NSDictionary *)dict {
    self = [super initWithDict:dict];
    if (self) {
        self.queryType = SFSyncDownTargetQueryTypeCustom;
        self.path = dict[kApexSyncDownTargetPath];
        self.queryParams = dict[kApexSyncDownTargetQueryParams];
        self.endpoint = dict[kApexSyncDownTargetEndpoint];
    }
    return self;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        self.queryType = SFSyncDownTargetQueryTypeCustom;
    }
    return self;
}

- (void)startFetch:(SFSmartSyncSyncManager*)syncManager maxTimeStamp:(long long)maxTimeStamp errorBlock:(SFSyncDownTargetFetchErrorBlock)errorBlock completeBlock:(SFSyncDownTargetFetchCompleteBlock)completeBlock {
    
    __weak ApexSyncDownTarget *weakSelf = self;
    
    SFRestRequest* request = [SFRestRequest requestWithMethod:SFRestMethodGET path:self.path queryParams:self.queryParams];
    [request setEndpoint:self.endpoint];
    
    [SFSmartSyncNetworkUtils sendRequestWithSmartSyncUserAgent:request failBlock:errorBlock completeBlock:^(id response) {
        if ([response isKindOfClass:[NSArray class]]) {
            NSArray *array = (NSArray *)response;
            weakSelf.totalSize = array.count;
            completeBlock(array);
        } else if ([response isKindOfClass:[NSDictionary class]]) {
            NSDictionary *dictionary = (NSDictionary *)response;
            if ([dictionary valueForKey:@"Id"] != (NSString *)[NSNull null]) {
                weakSelf.totalSize = 1;
                completeBlock(@[dictionary]);
            } else {
                NSError *error = [NSError errorWithDomain:@"apex" code:100 userInfo:@{@"msg" : @"Missing salesforce id"}];
                errorBlock(error);
            }
        } else {
            NSError *error = [NSError errorWithDomain:@"apex" code:100 userInfo:@{@"msg" : @"Invalid response type"}];
            errorBlock(error);
        }
    }];
}

- (NSMutableDictionary *)asDict {
    NSMutableDictionary *dict = [super asDict];
    dict[kSFSyncTargetiOSImplKey] = @"ApexSyncDownTarget";
    dict[kApexSyncDownTargetPath] = self.path;
    dict[kApexSyncDownTargetQueryParams] = self.queryParams;
    dict[kApexSyncDownTargetEndpoint]  = self.endpoint;
    return dict;
}

+ (ApexSyncDownTarget *)newSyncTarget:(NSString *)path queryParams:(NSDictionary *)queryParams {
    NSDictionary *dictionary = @{kApexSyncDownTargetPath : path, kApexSyncDownTargetQueryParams : queryParams};
    return [[ApexSyncDownTarget alloc] initWithDict:dictionary];
}

@end

File: CSMobileBase/Salesforce/ApexSyncDownTarget.swift
//
//  ApexSyncDownTarget.swift
//  Pods
//
//  Created by Mayank Bhayana on 4/21/18.
//  Copyright © 2018 Textron, Inc. All rights reserved.
//

import Foundation
import SmartSync

public let kApexSyncDownTargetId = "Id";
public let kApexSyncDownTargetPath = "path";
public let kApexSyncDownTargetQueryParams = "queryParams";
public let kApexSyncDownTargetEndpoint = "endpoint";

@objc(ApexSyncDownTarget)
class ApexSyncDownTarget: SFSyncDownTarget {
    
    private var path: String
    private var queryParams: [AnyHashable:Any]
    public var endpoint: String
    
    override public init!(dict: [AnyHashable : Any]!) {
        path = dict[kApexSyncDownTargetPath] as! String
        queryParams = (dict[kApexSyncDownTargetQueryParams] as! Dictionary?) ?? [AnyHashable:Any]()
        endpoint = dict[kApexSyncDownTargetEndpoint] as! String
        super.init(dict: dict)
        queryType = .custom
    }
    
    override func asDict() -> NSMutableDictionary {
        let dict: NSMutableDictionary = super.asDict()
        dict[kSFSyncTargetiOSImplKey] = String(describing: type(of: self))
        dict[kApexSyncDownTargetPath] = path
        dict[kApexSyncDownTargetQueryParams] = queryParams
        dict[kApexSyncDownTargetEndpoint] = endpoint
        return dict
    }
    
    static func newSyncTarget(_ path: String, endpoint: String? = nil, queryParams: [AnyHashable : Any]!) -> ApexSyncDownTarget {
        let dict: [AnyHashable : Any] = [kApexSyncDownTargetPath : path, kApexSyncDownTargetQueryParams : queryParams, kApexSyncDownTargetEndpoint : endpoint ?? ""]
        return ApexSyncDownTarget(dict: dict)
    }

    override func startFetch(_ syncManager: SFSmartSyncSyncManager!, maxTimeStamp: Int64, errorBlock: SFSyncDownTargetFetchErrorBlock!, complete completeBlock: SFSyncDownTargetFetchCompleteBlock!) {
        
        SFLogger.log(SFLogLevel.debug, msg: "ApexSyncDownTarget request")
        
        if let queryParams = self.queryParams as? [String: String] {
            let request = SFRestRequest(method: .GET, path: path, queryParams: queryParams)
            request.endpoint = endpoint
            
            SFSmartSyncNetworkUtils.sendRequest(withSmartSyncUserAgent: request, fail: { (error) in
                SFLogger.log(SFLogLevel.error, msg: "CSRestSyncDownFile request failed: \(String(describing: error?.localizedDescription))")
                errorBlock(error)
            }) { [weak self] (response) in
                if let array = response as? NSArray {
                    self?.totalSize = UInt(array.count);
                    completeBlock(array as? [Any]);
                    SFLogger.log(SFLogLevel.debug, msg: "CSRestSyncDownFile request success - ARRAY")
                } else if let dictionary = response as? Dictionary<AnyHashable, Any> {
                    if dictionary[kApexSyncDownTargetId] != nil {
                        self?.totalSize = 1
                        completeBlock([dictionary])
                    } else {
                        let error = NSError(domain:"apex", code:100, userInfo: ["msg" : "Missing Salesforce Id"])
                        errorBlock(error)
                    }
                } else {
                    let error = NSError(domain:"apex", code:100, userInfo: ["msg" : "Invalid Response Type"])
                    errorBlock(error)
                }
            }
        }
    }
}
